### PR TITLE
Adding proper timeouts.

### LINKED ISSUE #9013: active log streaming times out after 5 minutes
I noticed that if I'm following a pod log that is being actively written to it will timeout after 5 minutes.  However if the pod is not writing logs the request will not timeout.  In that case the request will remain open seemingly indefinitely.

I can change the behavior by modifying this line of code:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/server.go#L68

To make it happen faster you can set the value lower.  I've been testing with 1 minute.

Steps to reproduce:

1. Launch cluster
2. Create the nginx pod in `examples/pod.yaml`
3. In another terminal generate log output by hitting the pod: `while :; do curl 172.17.0.22; sleep 0.5; done` (you'll need to change the ip).
4. Watch the logs: `time curl -ivs --raw "http://127.0.0.1:8080/api/v1beta3/namespaces/default/pods/nginx/log?container=nginx&follow=true&previous=false"`

In this case whenever the WriteTimeout is reached curl will receive the zero sized chunk and the connection will be closed.  If you leave out step 3 curl will continue seemingly indefinitely.

This behavior definitely seems odd.  Is the usage of WriteTimeout correct here?  A quick search shows that others have been confused by the intention of WriteTimeout which is why I ask.

### PR DESCRIPTION
**Pull request by krousey**
An attempt to properly fix #9013 and #9180. 

cc @bgrant0607 @thockin

### CODE REVIEW

**Comment by krousey** [comment]
Ideally we should push these timeouts down to the specific handlers and not have it at the top level.

I also may be unnecessary supporting http.Hijacker and http.CloseNotifier. We may only need those on connections that aren't subject to timeouts any way. To support that, there was a combinatoric explosion of types creates (4 in total).

If we can eventually push timeouts down to individual handlers that need them, and those specific ones don't need close notification or hijacking, this can greatly simplify.

**Comment by krousey** [comment]
cc @dchen1107

**Comment by zmerlynn** [comment]
Assigning back to yourself. Assign the oncall when you take "WIP" off.

**Comment by thockin** [comment]
Thanks for digging into this one

**Comment by bgrant0607** [comment]
Since still WIP, this has missed the 1.0 window. Can go into the next release.

**Comment by krousey** [comment]
I have added tests.

@dchen1107 Do you think some timeout handling should be done in kubelet too?

File: pkg/apiserver/handlers.go [hunk 1]
```diff
@@ -135,6 +140,163 @@ func RecoverPanics(handler http.Handler) http.Handler {
...
+// the handler times out, writes by h to its http.ResponseWriter will return
+// http.ErrHandlerTimeout. If timeoutFunc returns a nil timeout channel, no
+// timeout will be enforced.
+func TimeoutHandler(h http.Handler, timeoutFunc func(*http.Request) (timeout <-chan time.Time, msg string)) http.Handler {
+	return &timeoutHandler{h, timeoutFunc}
+}
+
+type timeoutHandler struct {
+	handler http.Handler
+	timeout func(*http.Request) (<-chan time.Time, string)
```
Comment (brendandburns) [inline pkg/apiserver/handlers.go]: Document the expectations of this function.  (in particular that returning nil is allowed and cauases it to early terminate.

**Comment by brendandburns** [comment]
one small nit.  otherwise LGTM.

File: pkg/apiserver/handlers.go (hunk 1 above)
Comment (krousey) [inline pkg/apiserver/handlers.go]: Actually returning nil will not cause it to terminate early. A nil return will in effect be no timeout. And this is documented in that last sentence of the public function 7 lines above.

File: pkg/kubelet/server.go [hunk 2]
```diff
@@ -65,8 +65,6 @@ func ListenAndServeKubeletServer(host HostInterface, address net.IP, port uint,
 	s := &http.Server{
 		Addr:           net.JoinHostPort(address.String(), strconv.FormatUint(uint64(port), 10)),
 		Handler:        &handler,
-		ReadTimeout:    5 * time.Minute,
-		WriteTimeout:   5 * time.Minute,
```
Comment (liggitt) [inline pkg/kubelet/server.go]: this drops the timeouts on the kubelet server, but I don't see how the new TimeoutHandler is getting set up for the kubelet, only the API server... am I missing something?

File: pkg/kubelet/server.go [hunk 3]
```diff
@@ -86,8 +84,6 @@ func ListenAndServeKubeletReadOnlyServer(host HostInterface, address net.IP, por
 	server := &http.Server{
 		Addr:           net.JoinHostPort(address.String(), strconv.FormatUint(uint64(port), 10)),
 		Handler:        &s,
-		ReadTimeout:    5 * time.Minute,
-		WriteTimeout:   5 * time.Minute,
```
Comment (liggitt) [inline pkg/kubelet/server.go]: same here

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -415,13 +415,19 @@ func (s *APIServer) Run(_ []string) error {
 	}
 
 	longRunningRE := regexp.MustCompile(s.LongRunningRequestRE)
+	longRunningTimeout := func(req *http.Request) (<-chan time.Time, string) {
+		// TODO unify this with apiserver.MaxInFlightLimit
+		if longRunningRE.MatchString(req.URL.Path) || req.URL.Query().Get("watch") == "true" {
+			return nil, ""
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: does this make long-running requests of unlimited duration?

Comment (krousey) [inline cmd/kube-apiserver/app/server.go]: As long as the client is still connected.

File: pkg/kubelet/server.go (hunk 2 above)
Comment (krousey) [inline pkg/kubelet/server.go]: No. It is not set up. I was under the impression that the API calls were proxied through the API server, so they would handle the timeouts.

File: cmd/kube-apiserver/app/server.go (hunk 4 above)
Comment (krousey) [inline cmd/kube-apiserver/app/server.go]: I would also note that Read and Write timeouts on the server object don't actually close connections. They only cause errors on read and write. It's up to the error handling code to close it in that case. The bugs mentioned in this PR were two that actually held connections open for a long time, but didn't close until there was IO interaction to trigger it.

**Comment by xiang90** [comment]
@krousey You mentioned that we might have a few advanced use case. None of them actually happened over the last year. Can I assume that we actually do not need them? This implementation is too racy. Can we switch to use something simpler like https://golang.org/pkg/net/http/#TimeoutHandler or a slightly modified version of that?

**Comment by liggitt** [comment]
> This implementation is too racy

@xiang90 do you have specific issues you can reference?

> Can we switch to use something simpler like https://golang.org/pkg/net/http/#TimeoutHandler

That impl is simpler, but does not actually work with the mix of handlers we have (long running, chunk writing, streaming, etc). From the godoc:

> TimeoutHandler buffers all Handler writes to memory and does not support the Hijacker or Flusher interfaces.

I think any timeout handler that does work with those interfaces is likely to be more complex

**Comment by xiang90** [comment]
@liggitt 

> do you have specific issues you can reference?

https://github.com/kubernetes/kubernetes/issues/29001#issuecomment-234391645
https://github.com/kubernetes/kubernetes/issues/29001#issuecomment-234812082

> That impl is simpler, but does not actually work with the mix of handlers 

No. The timeout handler can wrap any handlers. You can have a timeout handler per handler.

**Comment by liggitt** [comment]
Did you read the comment about TimeoutHandler buffering all writes to memory, and not working with Hijacker interfaces?

**Comment by xiang90** [comment]
@liggitt 

Why you assume I did not? 

Can you please read https://github.com/kubernetes/kubernetes/pull/10656#issuecomment-117859234 and my comment at https://github.com/kubernetes/kubernetes/pull/10656#issuecomment-234812371? 

I am actually not sure what you are trying to explain. If you know there is a case that we need hijacker or flush, probably you should just post it here. Or if you think the current implement is not racy, you can just tell me why. If you want to fix the race, go ahead and do it. Thanks.

**Comment by ncdc** [comment]
We use hijack for exec, attach/run, and port forwarding currently.
Eventually I'd like to move to an HTTP/2 implementation so hijacking
wouldn't be required any more.

**Comment by xiang90** [comment]
@ncdc Hijacking and timeout cannot work well together I think. Once we hijack it, we have to cancel the timeout handler and set TCP timeout on the conn. The timeout handler cannot assume the inner hijacked handler speaks http anymore. Writing HTTP timeout headers once it timeouts makes no sense after the hijacking. That is part of the reason standard lib does not support timeout + hijacking i believe. 

I did `grep Hijacker -r .` and failed to find a case that we really need hijacker. Maybe I was wrong.

**Comment by ncdc** [comment]
@xiang90 see https://github.com/kubernetes/kubernetes/blob/ef0c9f0c5b8efbba948a0be2c98d9d2e32e0b68c/pkg/util/httpstream/spdy/upgrade.go#L54-L65

I agree we don't want to timeout a hijacked connection (the current spdy implementation for exec/attach/port-forward has its own idle timeout functionality).

**Comment by xiang90** [comment]
@ncdc Oh. I missed that one... Was assuming it is a dependency :(. OK. Then that makes sense. The spdy thing will go away once we use HTTP2 by default which does multiplexing on TCP? What is your suggestion for moving forward for now? also /cc @smarterclayton

**Comment by ncdc** [comment]
We will need time to engineer the HTTP2 replacement, plus we'll have to go through a deprecation period for the spdy implementation.

The go HTTP2 implementation allows you to read from/write to the request and response bodies without having to go through any sort of hijacking. We'll have to write our own multiplexing to support multiple "streams" (stdin, stdout, stderr, error, resize, etc), however.

**Comment by xiang90** [comment]
@ncdc OK. So we still need this. How do you think about cancelling http timeout once we hijacked the connection from timeout writer?

**Comment by ncdc** [comment]
@xiang90 there are a few routes that are currently excluded from the timeout handler, including logs, exec, attach, and portforward:

https://github.com/kubernetes/kubernetes/blob/411c32bc98cff616d80575819258a62f5df8bc86/pkg/genericapiserver/options/server_run_options.go#L46

https://github.com/kubernetes/kubernetes/blob/411c32bc98cff616d80575819258a62f5df8bc86/pkg/genericapiserver/genericapiserver.go#L675

https://github.com/kubernetes/kubernetes/blob/411c32bc98cff616d80575819258a62f5df8bc86/pkg/apiserver/handlers.go#L186

Can we continue to do this?

**Comment by xiang90** [comment]
@ncdc OK... So that means that the timeout handler does not need to support hijacking at all, since the one use hijacking does not under timeout handler?

**Comment by ncdc** [comment]
@xiang90 I think that's correct, but let's make sure @liggitt and @smarterclayton agree too 😄

**Comment by liggitt** [comment]
possible. I also wonder about the watch handling endpoints, which write chunked data and expect it to be flushed to the client, but also be able to timeout.

**Comment by smarterclayton** [comment]
Watch uses hijack for web sockets, but only needs to flush for chunked.  I
thought for watch we set timeouts on the connection ourselves.

**Comment by xiang90** [comment]
@liggitt @smarterclayton 

OK. So we do not need timeout for spdy upgrade as @ncdc mentioned. We do not need timeout for watch, which requires flush, since it should set timeout on connection. 

The we probably should just use the standard timeout handler (which does not support hijacking and flush for good reasons) instead of the customized one.

I can make the change if @smarterclayton @liggitt @ncdc all agree.

**Comment by liggitt** [comment]
@ncdc do you know whether our current timeout is exercised for proxy paths?

**Comment by liggitt** [comment]
I just want to be careful that we don't regress spdy connections (exec/attach), websocket connections (exec/attach/log/watch), or streaming connections (log, watch, proxy). I also wonder about the perf impact of buffering all responses in memory... some responses can get quite large, especially for controllers that list across all namespaces.

**Comment by ncdc** [comment]
> do you know whether our current timeout is exercised for proxy paths?

@liggitt which timeout in particular? for exec/attach/pf?

**Comment by xiang90** [comment]
@liggitt First, we need to make things correct. Then we can start taking care about performance, optimization. Now it is broken.

**Comment by liggitt** [comment]
@ncdc if the current timeout handler is being used in proxy paths, or if proxy gets exempted from timeouts

**Pull request merged**